### PR TITLE
[hlm-707]: Generic query builder with Select and Update queries

### PR DESCRIPTION
**Pull request by shivab-egov**
Created generic query builder. 
- Select query
- Update query
- Need to work on insert query.

### CODE REVIEW

File: digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/DefaultQueryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
+package org.digit.health.sync.repository;
+
+import lombok.extern.slf4j.Slf4j;
```
Comment (jayantp-egov) [inline digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/DefaultQueryBuilder.java:3]: org.digit.health.common.data.query.annotations
org.digit.health.common.data.query.builder

File: digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryFieldConditionChecker.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package org.digit.health.sync.repository;
+
+import java.lang.reflect.Field;
+import java.util.Optional;
+
+public interface QueryFieldConditionChecker {
```
Comment (jayantp-egov) [inline digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryFieldConditionChecker.java:6]: Rename to QueryFieldChecker.java

File: digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryFieldConditionChecker.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    boolean check(Field field, Object object) throws IllegalAccessException;
+
+    QueryFieldConditionChecker checkIfFieldIsNotNull = (field, object) -> Optional.ofNullable(field.get(object)).isPresent();
```
Comment (jayantp-egov) [inline digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryFieldConditionChecker.java:9]: Remove checkIf
Rename to isNotNull
Rename to isAnnotatedWithUpdateBy

**Review by jayantp-egov — CHANGES_REQUESTED** [review]

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/SelectQueryBuilder.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+    @Override
+    public String build(Object object) {
+        StringBuilder queryStringBuilder = null;
+        try {
+            String tableName = GenericQueryBuilder.getTableName(object.getClass());
+            List<String> whereClauses = GenericQueryBuilder.getFieldsWithCondition(object, QueryFieldChecker.isNotNull);
+            queryStringBuilder = GenericQueryBuilder.generateQuery(GenericQueryBuilder.selectQueryTemplate(tableName), whereClauses);
+        } catch (Exception exception) {
+            System.out.println(exception.getMessage());
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/SelectQueryBuilder.java:15]: Create exception QueryBuilderException in the query folder and throw it from here by putting exception.getMessage() in the constructor.
`throw new QueryBuilderException(exception.getMessage);`

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/UpdateQueryBuilder.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    @Override
+    public String build(Object object) {
+        StringBuilder queryStringBuilder = null;
+        try {
+            String tableName = GenericQueryBuilder.getTableName(object.getClass());
+            List<String> fieldsToUpdate = GenericQueryBuilder.getFieldsWithCondition(object,  QueryFieldChecker.isNotNull);
+            List<String> fieldsToUpdateWith = GenericQueryBuilder.getFieldsWithCondition(object,  QueryFieldChecker.isAnnotatedWithUpdateBy);
+            queryStringBuilder = GenericQueryBuilder.generateQuery(GenericQueryBuilder.updateQueryTemplate(tableName), fieldsToUpdate, fieldsToUpdateWith);
+        } catch (Exception exception) {
+            System.out.println(exception.getMessage());
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/UpdateQueryBuilder.java:15]: Do the same thing here as well.
`throw new QueryBuilderException(exception.getMessage);`

**Review by jayantp-egov — CHANGES_REQUESTED** [review]

File: digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryBuilder.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package org.digit.health.sync.repository;
+
+public interface QueryBuilder {
+    String buildSelectQuery(Object object);
```
[the remark this answers is not in the document]
Comment (shivab-egov) [reply digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryBuilder.java]: Done.

File: digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/DefaultQueryTemplate.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package org.digit.health.sync.repository;
+
+public class DefaultQueryTemplate {
+    public static String select(String tableName){
+        return String.format("SELECT * FROM %s", tableName);
```
[the remark this answers is not in the document]
Comment (shivab-egov) [reply digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/DefaultQueryTemplate.java]: Done.

File: digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryBuilderUtils.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.IntStream;
+
+public class QueryBuilderUtils {
+
+    public String getTableName(Class reflectClass){
```
[the remark this answers is not in the document]
Comment (shivab-egov) [reply digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryBuilderUtils.java]: Done.

File: digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/DefaultQueryBuilder.java (hunk 1 above)
Comment (shivab-egov) [reply digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/DefaultQueryBuilder.java]: Done.

File: digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryFieldConditionChecker.java (hunk 2 above)
Comment (shivab-egov) [reply digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryFieldConditionChecker.java]: Done.

File: digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryFieldConditionChecker.java (hunk 3 above)
Comment (shivab-egov) [reply digit-health-services/health-sync-service/src/main/java/org/digit/health/sync/repository/QueryFieldConditionChecker.java]: Done.

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/SelectQueryBuilder.java (hunk 4 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/SelectQueryBuilder.java]: Created QueryBuilderException.

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/UpdateQueryBuilder.java (hunk 5 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/UpdateQueryBuilder.java]: Created QueryBuilderException.

**Review by jayantp-egov — APPROVED** [review]

**Pull request merged**